cardano-cli/src/Cardano/CLI/EraBased/Run/CreateTestnetData.hs

```diff
@@ -0,0 +1,656 @@
+      -- {0 -> delegate-keys/delegate0/vrf.vkey, 1 -> delegate-keys/delegate1/vrf.vkey, ...}
+      delegateVrfKeys = mkPaths "delegate-keys" "delegate" "vrf.vkey"
+
+  forM_ [ 1 .. numGenesisKeys ] $ \index -> do
+    createGenesisKeys (genesisDir </> ("genesis" <> show index))
+    createDelegateKeys keyOutputFormat (outputDir </> "delegate-keys" </> ("delegate" <> show index))
+
+  -- UTxO keys
+  let utxoKeys = [outputDir </> "utxo-keys" </> ("utxo" <> show index) </> "utxo.vkey"
+                  | index <-  [ 1 .. numUtxoKeys ]]
+  forM_ [ 1 .. numUtxoKeys ] $ \index ->
+    createUtxoKeys $ outputDir </> "utxo-keys" </> ("utxo" <> show index)
+
+  let mayStakePoolRelays = Nothing -- TODO @smelc temporary?
+
+  -- Pools
+  poolParams <- forM [ 1 .. numPools ] $ \index -> do
+    let poolsDir = outputDir </> "pools-keys"
+        poolDir = poolsDir </> ("pool" <> show index)
+
+    createPoolCredentials keyOutputFormat poolDir Nothing
+    buildPoolParams network poolDir Nothing (fromMaybe mempty mayStakePoolRelays)
+
+  -- Stake delegators
+  let (delegsPerPool, delegsRemaining) = divMod numStakeDelegators numPools
+      delegsForPool poolIx = if delegsRemaining /= 0 && poolIx == numPools
+        then delegsPerPool
+        else delegsPerPool + delegsRemaining
+      distribution = [pool | (pool, poolIx) <- zip poolParams [1 ..], _ <- [1 .. delegsForPool poolIx]]
+
+  g <- Random.getStdGen
+
+  -- Distribute M delegates across N pools:
+  delegations <- liftIO $ Lazy.forStateM g distribution $ flip computeInsecureDelegation network
+
+  genDlgs <- readGenDelegsMap genesisVKeysPaths delegateKeys delegateVrfKeys
+  nonDelegAddrs <- readInitialFundAddresses utxoKeys network
+  start <- maybe (SystemStart <$> getCurrentTimePlus30) pure systemStart
+
+  stuffedUtxoAddrs <- liftIO $ Lazy.replicateM (fromIntegral numStuffedUtxo) genStuffedAddress
+
+  let stake = second Ledger.ppId . mkDelegationMapEntry <$> delegations
+      stakePools = [ (Ledger.ppId poolParams', poolParams') | poolParams' <- snd . mkDelegationMapEntry <$> delegations ]
+      delegAddrs = dInitialUtxoAddr <$> delegations
+      !shelleyGenesis' =
+        updateCreateStakedOutputTemplate
+          -- Shelley genesis parameters
+          start genDlgs supply (length nonDelegAddrs) nonDelegAddrs stakePools stake
+          supplyDelegated (length delegations) delegAddrs stuffedUtxoAddrs shelleyGenesis
+
+  -- Write genesis.json file to output
+  liftIO $ LBS.writeFile (outputDir </> "genesis.json") $ Aeson.encode shelleyGenesis'
+  where
+    genesisDir = outputDir </> "genesis-keys"
+    adjustTemplate t = t { sgNetworkMagic = unNetworkMagic (toNetworkMagic network) }
+    keyOutputFormat = KeyOutputFormatTextEnvelope
+    mkDelegationMapEntry :: Delegation -> (Ledger.KeyHash Ledger.Staking StandardCrypto, Ledger.PoolParams StandardCrypto)
+    mkDelegationMapEntry d = (dDelegStaking d, dPoolParams d)
+
+genStuffedAddress :: IO (AddressInEra ShelleyEra)
```

Comment:
Yeah that is correct, nevertheless I think it's a good name because it makes clear it handles the flag named `stuffed`, so I'm leaving it like this :+1: